nav2_costmap_2d/src/observation_buffer.cpp

```diff
@@ -171,40 +168,27 @@ void ObservationBuffer::bufferCloud(const sensor_msgs::msg::PointCloud2 & cloud)
   purgeStaleObservations();
 }
 
-// returns a copy of the observations
-void ObservationBuffer::getObservations(std::vector<Observation> & observations)
+// appends the observations from the buffer to the passed vector
+void ObservationBuffer::getObservations(std::vector<Observation::ConstSharedPtr> & observations)
 {
   // first... let's make sure that we don't have any stale observations
   purgeStaleObservations();
 
-  // now we'll just copy the observations for the caller
-  std::list<Observation>::iterator obs_it;
-  for (obs_it = observation_list_.begin(); obs_it != observation_list_.end(); ++obs_it) {
-    observations.push_back(*obs_it);
-  }
+  // now we'll just copy the observation ptrs for the caller
+  observations.insert(observations.end(), observation_list_.cbegin(), observation_list_.cend());
 }
 
 void ObservationBuffer::purgeStaleObservations()
 {
   if (!observation_list_.empty()) {
-    std::list<Observation>::iterator obs_it = observation_list_.begin();
     // if we're keeping observations for no time... then we'll only keep one observation
     if (observation_keep_time_ == rclcpp::Duration(0.0s)) {
-      observation_list_.erase(++obs_it, observation_list_.end());
-      return;
-    }
-
-    // otherwise... we'll have to loop through the observations to see which ones are stale
-    for (obs_it = observation_list_.begin(); obs_it != observation_list_.end(); ++obs_it) {
-      Observation & obs = *obs_it;
-      // check if the observation is out of date... and if it is,
-      // remove it and those that follow from the list
-      if ((clock_->now() - obs.cloud_->header.stamp) >
-        observation_keep_time_)
-      {
-        observation_list_.erase(obs_it, observation_list_.end());
-        return;
-      }
+      observation_list_.erase(std::next(observation_list_.begin()), observation_list_.end());
+    } else {
+      // otherwise remove all observations that are older then the keep time
+      observation_list_.remove_if([now = clock_->now(), this](const auto observation){
```

Comment:
The previous implementation was assuming that the `observation_list_` was ordered, but I think that is not guaranteed.